sysl2/parser.go

```diff
@@ -0,0 +1,329 @@
+package main
+
+import (
+	"fmt"
+
+	"anz-bank/sysl/sysl2/proto"
+)
+
+func makeQuantifierOptional() *sysl.Quantifier {
+	return &sysl.Quantifier{Union: &sysl.Quantifier_Optional{}}
+}
+
+func makeQuantifierZeroPlus() *sysl.Quantifier {
+	return &sysl.Quantifier{Union: &sysl.Quantifier_ZeroPlus{}}
+}
+
+func makeQuantifierOnePlus() *sysl.Quantifier {
+	return &sysl.Quantifier{Union: &sysl.Quantifier_OnePlus{}}
+}
+
+func makeStringTerm(str string) *sysl.Term {
+	return &sysl.Term{Atom: &sysl.Atom{Union: &sysl.Atom_String_{String_: str}}, Quantifier: nil}
+}
+
+func makeRegexpTerm(str string) *sysl.Term {
+	return &sysl.Term{Atom: &sysl.Atom{Union: &sysl.Atom_Regexp{Regexp: str}}, Quantifier: nil}
+}
+
+func makeSequence(terms ...*sysl.Term) *sysl.Sequence {
+	seq := sysl.Sequence{Term: terms}
+	return &seq
+}
+
+func makeRule(name string) (*sysl.RuleName, *sysl.Term) {
+	ruleName := sysl.RuleName{Name: name}
+	ruleTerm := sysl.Term{Atom: &sysl.Atom{Union: &sysl.Atom_Rulename{Rulename: &ruleName}}, Quantifier: nil}
+	return &ruleName, &ruleTerm
+}
+
+// S –> bab | bA
+// A –> d | cA
+
+func makeGrammar1() sysl.Grammar {
+	a := makeStringTerm("a")
+	b := makeStringTerm("b")
+	c := makeStringTerm("c")
+	d := makeStringTerm("d")
+
+	ruleNameA, A := makeRule("A")
+
+	a1 := makeSequence(d)
+	a2 := makeSequence(c, A)
+
+	choiceA := sysl.Choice{Sequence: []*sysl.Sequence{a1, a2}}
+	ruleA := sysl.Rule{Name: ruleNameA, Choices: &choiceA}
+
+	ruleNameS, _ := makeRule("S")
+	s1 := makeSequence(b, a, b)
+	s2 := makeSequence(b, A)
+	choiceS := sysl.Choice{Sequence: []*sysl.Sequence{s1, s2}}
+	ruleS := sysl.Rule{Name: ruleNameS, Choices: &choiceS}
+
+	prods := map[string]*sysl.Rule{}
+	prods["S"] = &ruleS
+	prods["A"] = &ruleA
+
+	return sysl.Grammar{Name: "test", Rules: prods}
```

Comment:
Make this function one big literal. Ditto for every make… function below.

Reply:
done.